src/client/Microsoft.Identity.Client/Cache/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,603 @@
+            //      and start a task again (though this scenario is unlikely to happen)
+            //
+            // Without the EventQueueTaskDoNotStop state check below, steps (3), (4) and (5) above will not be applicable.
+            // After step (2) the event queue task is still running and the state is still EventQueueTaskRunning (even though the EventQueueTaskAction() method has already checked that the queue is empty
+            // and is about to stop the task). This method (StartEventQueueTaskIfNotRunning()) will return, the execution will switch over to EventQueueTaskAction(),
+            // and the task will terminate. This means no new task would be started to process the newly added event.
+            //
+            // This scenario is unlikely to happen, as it can only occur if the event queue task ALREADY checked the queue and it was empty, and the new event was added AFTER that check but BEFORE the
+            // event queue task set the _eventQueueTaskState to EventQueueTaskStopped.
+
+            if (Interlocked.CompareExchange(ref _eventQueueTaskState, EventQueueTaskDoNotStop, EventQueueTaskRunning) == EventQueueTaskRunning)
+            {
+                return;
+            }
+
+            // If the task is stopped, set _eventQueueTaskState = EventQueueTaskRunning and start a new task.
+            // Note: we need to call the Task.Run() to start a new task on the default TaskScheduler (TaskScheduler.Default) so it does not interfere with
+            // the caller's TaskScheduler (if there is one) as some custom TaskSchedulers might be single-threaded and its execution can be blocked.
+            if (Interlocked.CompareExchange(ref _eventQueueTaskState, EventQueueTaskRunning, EventQueueTaskStopped) == EventQueueTaskStopped)
+            {
+                Task.Run(EventQueueTaskAction);
+            }
+        }
+
+        /// Each time a node gets accessed, it gets moved to the beginning (head) of the list if the _maintainLRU == true
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            if (!_map.TryGetValue(key, out var cacheItem))
+            {
+                value = default;
+                return false;
+            }
+
+            // make sure node hasn't been removed by a different thread
+            if (_maintainLRU)
+            {
+                AddActionToEventQueue(() =>
+                {
+                    _doubleLinkedList.Remove(cacheItem);
+                    _doubleLinkedList.AddFirst(cacheItem);
+                });
+            }
+
+            value = cacheItem != null ? cacheItem.Value : default;
+            return cacheItem != null;
+        }
+
+        public IList<TValue> GetAllValues()
```

Comment:
@dannybtsai Was there any talks about adding a GetAllValues method (kind of like this) that returns actual values and not just CacheEntries?

Reply:
@pmaytak that could be expensive as we have seen 100,000+ items in the cache.